src/Types/Workspace/Creator.php

```diff
@@ -3,36 +3,116 @@
+        file_put_contents($dir . '/workspace.override.yml', $override);
+    }
+
+    private function toYamlLooseKey(string $key, mixed $data): string
+    {
+        // workspace magic keys aren't to be quoted, so just yaml dump it's value
+        return $key . ":\n" . preg_replace('/^/m', '  ', Yaml::dump($data, 10, 2));
+    }
+
+    private function findHarnessLayers(string $harness): array
+    {
+        $package = $this->packages->get($harness);
+
+        $harnessLayers = [$harness];
+
+        $harnessData = $this->parseYamlMergeStreams($this->downloadAndExtractHarnessYml($package));
+        if (!is_array($harnessData)) {
+            throw new \Exception('Could not parse the harness\'s harness.yml file');
         }
 
-        $workspace[] = '';
+        foreach ($harnessData as $key => $value) {
+            if (str_starts_with($key, 'harness(')) {
+                if (isset($value['parentLayers'])) {
+                    if (!is_array($value['parentLayers'])) {
+                        throw new \Exception('Could not parse the harness\'s harness.yml file');
+                    }
+                    $harnessLayers = array_merge($value['parentLayers'], $harnessLayers);
+                }
+                break;
+            }
+        }
 
-        $override = [];
-        $override[] = "key('default'): " . (new Key('default'))->getKeyAsHex();
-        $override[] = '';
+        return $harnessLayers;
+    }
+
+    private function parseYamlMergeStreams($content): array
+    {
+        // Workspace doesn't follow Yaml spec and merges Yaml documents together
+        // Symfony Yaml doesn't support Yaml document boundaries
+        $documents = preg_split('/\R---\R/', $content);
+        $mergedDocument = [];
+        foreach ($documents as $document) {
+            $declarations = Yaml::parse($document);
+            if (!is_array($declarations)) {
+                continue;
+            }
+            $mergedDocument = array_merge($mergedDocument, $declarations);
+        }
+
+        return $mergedDocument;
+    }
+
+    private function downloadAndExtractHarnessYml(Package $package): string
+    {
+        $packageTarball = tempnam(sys_get_temp_dir(), 'my127ws');
+        file_put_contents($packageTarball, file_get_contents($package->getDist()['url']));
+
+        $packageDir = tempnam(sys_get_temp_dir(), 'my127ws');
+        unlink($packageDir);
+        if (!mkdir($packageDir, 0700)) {
+            throw new \Exception('Could not create temporary ' . $packageDir . ' directory for harness');
+        }
+
+        $command = 'tar -zxf ' . escapeshellarg($packageTarball) . ' --strip=1 -C ' . escapeshellarg($packageDir);
+        if (passthru($command) === false) {
+            throw new \Exception('Could not parse the harness\'s harness.yml file');
+        }
+
+        $harnessYaml = file_get_contents($packageDir . '/harness.yml');
+        exec('rm -rf ' . escapeshellarg($packageDir));
```

Comment:
This scares me somewhat

Reply:
I just wish PHP was another language where I could just import a archive system that can read the harness.yml directly from tarball stream

Reply:
PharData can do that, as with the tests but as can see nasty global storage it uses

Reply:
I'd have done that but I don't like that PharData can't mime detect file type, only read based on file extension, which is messy

Reply:
would the suggestions help?

Reply:
Sorry it was more the `rm -rf` aspect of it. It's likely safe but I get scared because of the incidents with Bumblebee & Steam for linux :)